### PR TITLE
Refactor/backend 2

### PR DESCRIPTION
**Pull request by ale8k**
Pretty chunky merge, complete refactor and ts implementation. Follows SOLID, DRY etc., and a couple other principles. 

Should be easily extensible now.

As discussed with Ken, testing will be it's own branch.

### CODE REVIEW

File: server/src/socketio-events/scrum-poker/constants/EVENT_CONSTANTS.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export const USER_JOIN = "join";
+export const USER_SEND_ESTIMATE = "sendEstimate";
+export const USER_VIEW_ESTIMATES = "clickExpand";
+export const USER_DISCONNECT = "disconnect";
+
+
```
Comment (ale8k) [inline server/src/socketio-events/scrum-poker/constants/EVENT_CONSTANTS.ts:6]: Excess lines, will remove.

File: server/src/socketio-events/scrum-poker/events/UserDisconnect.sio.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * @class
+ * @event UserDisconnect#constructor:disconnect
+ */
+export default class UserDisconnect extends SocketIOEvent {
+    constructor(io: socketio.Server , socket: Socket) {
+        super(USER_DISCONNECT, () => {
+            const room = UserHandler.getUserBySocketId(socket.id)?.room as string;
+
+            if (UserHandler.removeUserFromLocalStore(socket.id)) {
```
Comment (ale8k) [inline server/src/socketio-events/scrum-poker/events/UserDisconnect.sio.ts:17]: Perhaps this block may need documenting? I.e., why we broadcast back and specifically call 'true'? Idk if it's obvious the underlying connection stays alive.

File: server/src/socketio-events/scrum-poker/events/UserJoin.sio.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * The primary event for joining a room with a predetermined estimate
+ *
+ * @class
+ * @event Join#constructor:join
+ */
+export default class UserJoin extends SocketIOEvent {
+    constructor(io: socketio.Server , socket: Socket) {
+        super(USER_JOIN, ({ users_name, room, estimate }: IUser, acknowledgeFn) => {
+            UserHandler.addUserToLocalStore(new User(
+                users_name.trim().toLowerCase(),
```
Comment (ale8k) [inline server/src/socketio-events/scrum-poker/events/UserJoin.sio.ts:18]: Maybe the frontend could do this trim logic etc? Seems a little redundant and off-method-topic, should we worry the data sent may be incorrectly formatted?

Comment (KenErasga) [reply server/src/socketio-events/scrum-poker/events/UserJoin.sio.ts:18]: Yeah, I was going to get rid of it before. I don't think that's needed there. We also wanted some capitalise letters for their name when we get it back to the frontend.

File: server/src/socketio-events/scrum-poker/events/UserDisconnect.sio.ts (hunk 2 above)
Comment (KenErasga) [reply server/src/socketio-events/scrum-poker/events/UserDisconnect.sio.ts:17]: it's up to you. I can understand it, so if you want you can.

**Pull request merged**